cmd/metal-api/internal/fsm/fsm.go

```diff
@@ -0,0 +1,303 @@
+	"enter_" + metal.ProvisioningEventWaiting.String():          handleStateTransition,
+	"enter_" + metal.ProvisioningEventInstalling.String():       handleStateTransition,
+	"enter_" + metal.ProvisioningEventBootingNewKernel.String(): handleStateTransition,
+	"enter_" + metal.ProvisioningEventMachineReclaim.String():   handleStateTransition,
+
+	"enter_" + metal.ProvisioningEventPXEBooting.String():    resetFailedReclaim,
+	"enter_" + metal.ProvisioningEventPreparing.String():     resetFailedReclaim,
+	"enter_" + metal.ProvisioningEventPlannedReboot.String(): handlePlannedRebootEvent,
+	"enter_" + metal.ProvisioningEventPhonedHome.String():    handlePhonedHomeEvent,
+}
+
+func handleStateTransition(e *fsm.Event) {
+	provisioningFSM, event, err := parseCallbackArguments(e.Args)
+	if err != nil {
+		e.Err = err
+		return
+	}
+
+	provisioningFSM.container.Events = append(provisioningFSM.container.Events, *event)
+	provisioningFSM.container.LastEventTime = &(*event).Time
+	provisioningFSM.container.Liveliness = metal.MachineLivelinessAlive
+}
+
+func resetFailedReclaim(e *fsm.Event) {
+	provisioningFSM, _, err := parseCallbackArguments(e.Args)
+	if err != nil {
+		e.Err = err
+		return
+	}
+
+	provisioningFSM.container.FailedMachineReclaim = false
+
+	handleStateTransition(e)
+}
+
+func handlePlannedRebootEvent(e *fsm.Event) {
+	provisioningFSM, _, err := parseCallbackArguments(e.Args)
+	if err != nil {
+		e.Err = err
+		return
+	}
+
+	provisioningFSM.container.CrashLoop = false
+
+	handleStateTransition(e)
+}
+
+func handlePhonedHomeEvent(e *fsm.Event) {
+	provisioningFSM, event, err := parseCallbackArguments(e.Args)
+	if err != nil {
+		e.Err = err
+		return
+	}
+
+	if e.Src != metal.ProvisioningEventMachineReclaim.String() {
+		handleStateTransition(e)
+	} else if event.Time.Sub(*provisioningFSM.container.LastEventTime) > timeOutAfterMachineReclaim {
+		provisioningFSM.container.LastEventTime = &event.Time
+		provisioningFSM.container.FailedMachineReclaim = true
+	}
+
+	provisioningFSM.container.CrashLoop = false
+}
+
+func parseCallbackArguments(args []interface{}) (*ProvisioningFSM, *metal.ProvisioningEvent, error) {
+	if len(args) != 2 {
+		return nil, nil, errors.New("expecting two arguments of types *ProvisioningFSM, *metal.ProvisioningEvent")
+	}
+
+	provisioningFSM, ok := args[0].(*ProvisioningFSM)
+	if !ok {
+		return nil, nil, errors.New("first argument must be of type *ProvisioningFSM")
+	}
+
+	event, ok := args[1].(*metal.ProvisioningEvent)
+	if !ok {
+		return nil, nil, errors.New("second argument must be of type *metal.ProvisioningEvent")
+	}
+
+	return provisioningFSM, event, nil
+}
+
+func newProvisioningFSM(initialState metal.ProvisioningEventType, container *metal.ProvisioningEventContainer) *ProvisioningFSM {
+	f := fsm.NewFSM(
+		initialState.String(),
+		events,
+		callbacks,
+	)
+
+	provisioningFSM := ProvisioningFSM{
+		fsm:       f,
+		container: container,
+	}
+
+	return &provisioningFSM
```

Comment:
When returning this directly, you also do it inline:

```suggestion
	return &ProvisioningFSM{
		fsm:       f,
		container: container,
	}
```